src/object_store/src/object/opendal_engine.rs

```diff
@@ -0,0 +1,433 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Cursor;
+
+use bytes::{BufMut, Bytes, BytesMut};
+use fail::fail_point;
+use futures::future::try_join_all;
+use futures::StreamExt;
+use itertools::Itertools;
+use opendal::services::{hdfs, memory};
+use opendal::Operator;
+use tokio::io::AsyncRead;
+
+use super::{
+    BlockLocation, BoxedStreamingUploader, ObjectError, ObjectMetadata, ObjectResult, ObjectStore,
+    StreamingUploader,
+};
+/// Opendal object storage.
+#[derive(Clone)]
+pub struct OpendalObjectStore {
+    op: Operator,
+    engine_type: EngineType,
+}
+#[derive(Clone)]
+enum EngineType {
+    Memory,
+    Hdfs,
+}
+
+impl OpendalObjectStore {
+    /// create opendal hdfs engine.
+    pub fn new_hdfs_engine(namenode: String, root: String) -> Self {
+        // Create fs backend builder.
+        let mut builder = hdfs::Builder::default();
+        // Set the name node for hdfs.
+        builder.name_node(&namenode);
+        // Set the root for hdfs, all operations will happen under this root.
+        //
+        // NOTE: the root must be absolute path.
+        builder.root(&root);
+
+        // `Accessor` provides the low level APIs, we will use `Operator` normally.
+        let op: Operator = Operator::new(builder.build().unwrap());
+        Self {
+            op,
+            engine_type: EngineType::Hdfs,
+        }
+    }
+
+    /// create opendal memory engine, used for unit tests.
+    pub fn new_memory_engine() -> Self {
+        // Create fs backend builder.
+        let mut builder = memory::Builder::default();
+
+        let op: Operator = Operator::new(builder.build().unwrap());
+        Self {
+            op,
+            engine_type: EngineType::Memory,
+        }
+    }
+
+    async fn get_object<R, F>(&self, path: &str, f: F) -> ObjectResult<R>
+    where
+        F: Fn(&Bytes) -> R,
+    {
+        Some(&Bytes::from(self.op.object(path).read().await?))
+            .ok_or_else(|| ObjectError::internal(format!("no object at path '{}'", path)))
+            .map(f)
+    }
+}
+
+#[async_trait::async_trait]
+impl ObjectStore for OpendalObjectStore {
+    fn get_object_prefix(&self, _obj_id: u64) -> String {
+        String::default()
+    }
+
+    async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
+        if obj.is_empty() {
+            Err(ObjectError::internal("upload empty object"))
+        } else {
+            self.op.object(path).write(obj).await?;
+            Ok(())
+        }
+    }
+
+    fn streaming_upload(&self, path: &str) -> ObjectResult<BoxedStreamingUploader> {
```

Comment:
As discussed offline, we can look in to the `write_from` API to see whether we can easily support streaming write. `write_from` takes `AsyncRead`, which can be a async stream so it seems possible to push bytes to the stream while writing. I am okay with merging the current implementation first.

Reference:
- https://docs.rs/opendal/latest/opendal/struct.Object.html#method.write_from
- https://github.com/datafuselabs/opendal/blob/main/docs/rfcs/0554-write-refactor.md